diff --git a/include/circle/usb/lan7800.h b/include/circle/usb/lan7800.h
index e4623569..7b51ae83 100644
--- a/include/circle/usb/lan7800.h
+++ b/include/circle/usb/lan7800.h
@@ -45,6 +45,7 @@ public:
 
 	// returns TRUE if PHY link is up
 	boolean IsLinkUp (void);
+	boolean UpdatePHY (void);
 	
 	TNetDeviceSpeed GetLinkSpeed (void);
 
@@ -75,6 +76,7 @@ private:
 	CMACAddress m_MACAddress;
 
 	u32 m_FilterTable[33][2];
+	volatile boolean m_bLinkUp;
 };
 
 #endif
diff --git a/include/circle/usb/smsc951x.h b/include/circle/usb/smsc951x.h
index b1eb17ac..dec1ecdd 100644
--- a/include/circle/usb/smsc951x.h
+++ b/include/circle/usb/smsc951x.h
@@ -44,6 +44,7 @@ public:
 	
 	// returns TRUE if PHY link is up
 	boolean IsLinkUp (void);
+	boolean UpdatePHY (void);
 
 	TNetDeviceSpeed GetLinkSpeed (void);
 
@@ -69,6 +70,7 @@ private:
 	CUSBEndpoint *m_pEndpointBulkOut;
 
 	CMACAddress m_MACAddress;
+	volatile boolean m_bLinkUp;
 };
 
 #endif
diff --git a/lib/usb/lan7800.cpp b/lib/usb/lan7800.cpp
index 15250864..2274018b 100644
--- a/lib/usb/lan7800.cpp
+++ b/lib/usb/lan7800.cpp
@@ -232,7 +232,8 @@ static const char FromLAN7800[] = "lan7800";
 CLAN7800Device::CLAN7800Device (CUSBFunction *pFunction)
 :	CUSBFunction (pFunction),
 	m_pEndpointBulkIn (0),
-	m_pEndpointBulkOut (0)
+	m_pEndpointBulkOut (0),
+	m_bLinkUp (FALSE)
 {
 }
 
@@ -445,6 +446,11 @@ boolean CLAN7800Device::SendFrame (const void *pBuffer, unsigned nLength)
 
 boolean CLAN7800Device::ReceiveFrame (void *pBuffer, unsigned *pResultLength)
 {
+	if (!m_bLinkUp)
+	{
+		return FALSE;
+	}
+
 	assert (m_pEndpointBulkIn != 0);
 	assert (pBuffer != 0);
 	CUSBRequest URB (m_pEndpointBulkIn, pBuffer, FRAME_BUFFER_SIZE);
@@ -488,14 +494,21 @@ boolean CLAN7800Device::ReceiveFrame (void *pBuffer, unsigned *pResultLength)
 }
 
 boolean CLAN7800Device::IsLinkUp (void)
+{
+	return m_bLinkUp;
+}
+
+boolean CLAN7800Device::UpdatePHY (void)
 {
 	u16 usPHYModeStatus;
 	if (!PHYRead (0x01, &usPHYModeStatus))
 	{
-		return FALSE;
+		return TRUE;
 	}
 
-	return usPHYModeStatus & (1 << 2) ? TRUE : FALSE;
+	m_bLinkUp = usPHYModeStatus & (1 << 2) ? TRUE : FALSE;
+
+	return TRUE;
 }
 
 TNetDeviceSpeed CLAN7800Device::GetLinkSpeed (void)
diff --git a/lib/usb/smsc951x.cpp b/lib/usb/smsc951x.cpp
index 7dd6ac48..02cc2380 100644
--- a/lib/usb/smsc951x.cpp
+++ b/lib/usb/smsc951x.cpp
@@ -130,7 +130,8 @@ static const char FromSMSC951x[] = "smsc951x";
 CSMSC951xDevice::CSMSC951xDevice (CUSBFunction *pFunction)
 :	CUSBFunction (pFunction),
 	m_pEndpointBulkIn (0),
-	m_pEndpointBulkOut (0)
+	m_pEndpointBulkOut (0),
+	m_bLinkUp (FALSE)
 {
 }
 
@@ -274,6 +275,11 @@ boolean CSMSC951xDevice::SendFrame (const void *pBuffer, unsigned nLength)
 
 boolean CSMSC951xDevice::ReceiveFrame (void *pBuffer, unsigned *pResultLength)
 {
+	if (!m_bLinkUp)
+	{
+		return FALSE;
+	}
+
 	assert (m_pEndpointBulkIn != 0);
 	assert (pBuffer != 0);
 	CUSBRequest URB (m_pEndpointBulkIn, pBuffer, FRAME_BUFFER_SIZE);
@@ -317,14 +323,21 @@ boolean CSMSC951xDevice::ReceiveFrame (void *pBuffer, unsigned *pResultLength)
 }
 
 boolean CSMSC951xDevice::IsLinkUp (void)
+{
+	return m_bLinkUp;
+}
+
+boolean CSMSC951xDevice::UpdatePHY (void)
 {
 	u16 usPHYModeStatus;
 	if (!PHYRead (0x01, &usPHYModeStatus))
 	{
-		return FALSE;
+		return TRUE;
 	}
 
-	return usPHYModeStatus & (1 << 2) ? TRUE : FALSE;
+	m_bLinkUp = usPHYModeStatus & (1 << 2) ? TRUE : FALSE;
+
+	return TRUE;
 }
 
 TNetDeviceSpeed CSMSC951xDevice::GetLinkSpeed (void)
